sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -523,6 +532,14 @@ object SimplifyConditionals extends Rule[LogicalPlan] with PredicateHelper {
         } else {
           e.copy(branches = branches.take(i).map(branch => (branch._1, elseValue)))
         }
+
+      case EqualTo(i @ If(_, trueValue, falseValue), right: Literal)
+          if i.deterministic && isAlwaysFalse(trueValue :: falseValue :: Nil, right) =>
+        FalseLiteral
+
+      case EqualTo(c @ CaseWhen(branches, elseValue), right: Literal)
```

Comment:
A more general idea is to push down `EqualTo`(or other binary expressions) through If/CaseWhen.

For the example case, `(CASE WHEN id = 1 THEN 'b' WHEN id = 3 THEN 'c' END) = 'a'`, we go through the following steps
1. CASE WHEN id = 1 THEN 'b' = 'a' WHEN id = 3 THEN 'c' = 'a' END
2. CASE WHEN id = 1 THEN false WHEN id = 3 THEN false END
3. false

Step 3 may need more work in `SimplifyConditionals` to make it possible.

Reply:
+1 for this more general idea.

Reply:
We can create a new rule to do it, e.g. `PushFoldableIntoBranches`

Reply:
OK.